File: Data Structure/Array or Matrix/Medium/LET_no525_Contiguous Array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        cur, res = 0, 0
        record = dict()
        for i in range(len(nums)):
            if nums[i] == 0:
                cur -= 1
            else:
                cur += 1
            
            if cur == 0:
                res = max(res, i + 1)
            else:
                if cur in record:
                    res = max(res, i - record[cur])
                else:
                    record[cur] = i
        return res

File: Data Structure/Hash/Medium/LET_no525_Contiguous Array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int res = 0, cur = 0;
        unordered_map<int, int> memo;
        for (int i = 0; i < nums.size(); i++) {
            cur += nums[i] == 1 ? 1 : -1;
            if (cur == 0)
                res = max(res, i + 1);
            else {
                if (memo.count(cur))
                    res = max(res, i - memo[cur]);
                else
                    memo[cur] = i;
            }
        }

        return res;
    }
};